### ISSUE TITLE
B_박성빈_lab_week4

### ISSUE THREAD

**Issue by 03psb** [vlab-kaist/NeuralNet101#31]
**LAB 4 - 박성빈**

### Problem 1

'''
    import torch
    import torchvision.datasets as dsets
    import torchvision.transforms as transforms
    from torch.utils.data import DataLoader
    import torch.nn as nn

    class Softmax(nn.Module):
        def __init__(self):
            super().__init__()
            self.linear = nn.Linear(784, 10, bias=True)
            self.softmax = nn.Softmax(dim=1)

        def forward(self, x):
            x = self.linear(x)
            x = self.softmax(x)
            return x

    def main():
        num_epochs = 5
        batch_size = 100

        mnist_train = dsets.MNIST(root="MNIST_data/", train=True, transform=transforms.ToTensor(), download=True)
        data_loader = DataLoader(dataset=mnist_train, batch_size=batch_size, shuffle=True, drop_last=True)

        model = Softmax()
        optimizer = torch.optim.SGD(model.parameters(), lr=0.1)

        for epoch in range(num_epochs):
            for X, Y in data_loader:
                optimizer.zero_grad()
                hypothesis = model.forward(X.reshape(100, -1))
                cost = nn.CrossEntropyLoss()(hypothesis, Y)
                cost.backward()
                optimizer.step()
                print('Epoch : %d, cost = %.9f' % (epoch+1, cost))

        return model

    if __name__ == "__main__":
        main()
'''

### Problem 2

'''
    import torch.nn as nn

    class Softmax(nn.Module):
        def __init__(self):
            super().__init__()
            self.linear = nn.Linear(4, 3, bias=True)
            self.softmax = nn.Softmax(dim=1)

        def forward(self, x):
            x = self.linear(x)
            x = self.softmax(x)
            return x

    def main():
        num_epochs = 500
        filename = "Iris.csv"
        info = {"Iris-setosa":[1,0,0], "Iris-versicolor":[0,1,0], "Iris-virginica":[0,0,1]}

        df = pd.read_csv(filename)
        x_train = torch.tensor(df.loc[:, ['SepalLengthCm','SepalWidthCm','PetalLengthCm','PetalWidthCm']].values.tolist(), 
    dtype=torch.float32)
        y_train = torch.tensor([info[i] for i in df['Species'].values.tolist()], dtype=torch.float32)

        model = Softmax()
        optimizer = torch.optim.SGD(model.parameters(), lr=0.1)

        for epoch in range(num_epochs):
            optimizer.zero_grad()
            print(x_train.shape)
            print(y_train.shape)
            hypothesis = model.forward(x_train)
            cost = nn.CrossEntropyLoss()(hypothesis, y_train)
            cost.backward()
            optimizer.step()
            print('Epoch : %d, cost = %.9f' % (epoch + 1, cost))

        return model

    if __name__ == "__main__":
        main()
'''
